File: java/VendingMachine/src/com/ravendarque/vendingMachine/rails/Rail.java
package com.ravendarque.vendingMachine.rails;

public class Rail {

    private final double railPrice;
    private final String label;
    private int inventory;

    public Rail(int initialInventory, double price, String label) {

        validatePrice(price);

        this.inventory = initialInventory;
        this.railPrice = price;
        this.label = label;
    }

    public String getLabel() {

        return label;
    }

    public double getPrice() {

        return railPrice;
    }

    public boolean isEmpty() {

        return inventory == 0;
    }

    public void vend()
            throws RailEmptyException {

        if (isEmpty())
            throw new RailEmptyException();

        inventory--;
    }

    private static void validatePrice(double price) {

        if (price < 0)
            throw new IllegalArgumentException("Price cannot be negative");

    }
}

File: C++/VendingMachine/VendingMachine/Rail.cpp
#include "pch.h"
#include "Rail.h"
#include <utility>
#include "RailEmptyError.h"

Rail::Rail(const unsigned int initialInventory, const double price, std::string label):
	Inventory(initialInventory), Price(price), Label(std::move(label))
{
	ValidatePrice(price);
}

std::string Rail::GetLabel() const
{
	return Label;
}

double Rail::GetPrice() const
{
	return Price;
}

bool Rail::IsEmpty() const
{
	return Inventory == 0;
}

void Rail::Vend()
{
	if (IsEmpty())
		throw RailEmptyError();

	Inventory--;
}

bool Rail::CanVend() const
{
	return Inventory > 0;
}

void Rail::ValidatePrice(const double price)
{
	if (price < 0)
		throw std::invalid_argument("Price cannot be negative");
}

